core/src/main/scala/akka/persistence/r2dbc/state/scaladsl/DurableStateDao.scala

```diff
@@ -202,61 +329,121 @@ private[r2dbc] class DurableStateDao(settings: R2dbcSettings, connectionFactory:
           }
+
+        changeHandlers.get(entityType) match {
+          case None =>
+            recoverDataIntegrityViolation(r2dbcExecutor.updateOne(s"insert [${state.persistenceId}]")(insertStatement))
+          case Some(handler) =>
+            r2dbcExecutor.withConnection(s"insert [${state.persistenceId}]") { connection =>
+              val session = new R2dbcSession(connection)
+              for {
+                updatedRows <- recoverDataIntegrityViolation(R2dbcExecutor.updateOneInTx(insertStatement(connection)))
+                _ <- handler.process(session, change)
+              } yield updatedRows
+            }
+        }
       } else {
         val previousRevision = state.revision - 1
 
-        r2dbcExecutor.updateOne(s"update [${state.persistenceId}]") { connection =>
+        def updateStatement(connection: Connection): Statement = {
           val stmt = connection
-            .createStatement(updateStateSql(updateTags = true))
-            .bind(0, state.revision)
-            .bind(1, state.serId)
-            .bind(2, state.serManifest)
-            .bind(3, state.payload.getOrElse(Array.emptyByteArray))
-          bindTags(stmt, 4)
+            .createStatement(updateStateSql(entityType, updateTags = true, additionalBindings))
+            .bind(getAndIncIndex(), state.revision)
+            .bind(getAndIncIndex(), state.serId)
+            .bind(getAndIncIndex(), state.serManifest)
+            .bind(getAndIncIndex(), state.payload.getOrElse(Array.emptyByteArray))
+          bindTags(stmt, getAndIncIndex())
+          bindAdditionalColumns(stmt, additionalBindings)
 
           if (settings.dbTimestampMonotonicIncreasing) {
             if (settings.durableStateAssertSingleWriter)
               stmt
-                .bind(5, state.persistenceId)
-                .bind(6, previousRevision)
+                .bind(getAndIncIndex(), state.persistenceId)
+                .bind(getAndIncIndex(), previousRevision)
             else
               stmt
-                .bind(5, state.persistenceId)
+                .bind(getAndIncIndex(), state.persistenceId)
           } else {
             stmt
-              .bind(5, state.persistenceId)
-              .bind(6, previousRevision)
-              .bind(7, state.persistenceId)
+              .bind(getAndIncIndex(), state.persistenceId)
+              .bind(getAndIncIndex(), previousRevision)
+              .bind(getAndIncIndex(), state.persistenceId)
 
             if (settings.durableStateAssertSingleWriter)
-              stmt.bind(8, previousRevision)
+              stmt.bind(getAndIncIndex(), previousRevision)
             else
               stmt
           }
         }
+
+        changeHandlers.get(entityType) match {
+          case None =>
+            r2dbcExecutor.updateOne(s"update [${state.persistenceId}]")(updateStatement)
+          case Some(handler) =>
+            r2dbcExecutor.withConnection(s"update [${state.persistenceId}]") { connection =>
```

Comment:
```suggestion
            r2dbcExecutor.withConnection(s"update [${state.persistenceId}] (with change handler)") { connection =>
```
Or something hinting about that it's not just the regular update?

Reply:
added in 9f7c9de